pyk/src/pyk/cli/cli.py

```diff
@@ -1,8 +1,377 @@
+
+    def get_and_exec_command(self) -> None:
+        parser = self.create_argument_parser()
+        args = parser.parse_args()
+        stripped_args = {key: val for (key, val) in vars(args).items() if val != 'NoDefault'}
+        cmd = self.get_command(stripped_args)
+        cmd._options_group.extract(stripped_args)
+        cmd.exec()
+
+
+class Option:
+    _name: str
+    _aliases: list[str]
+    _cmd_line_name: str
+    _toml_name: str
+    _optional: bool
+    _help_str: str | None
+    _metavar: str | None
+    _default: Any
+
+    def __init__(
+        self,
+        name: str,
+        aliases: Iterable[str] = (),
+        cmd_line_name: str | None = None,
+        toml_name: str | None = None,
+        optional: bool = False,
+        help_str: str | None = None,
+        metavar: str | None = None,
+        default: Any = 'NoDefault',
+    ) -> None:
+        self._name = name
+        self._aliases = list(aliases)
+        self._cmd_line_name = cmd_line_name or name
+        self._toml_name = cmd_line_name or name
+        self._optional = optional
+        self._help_str = help_str
+        self._metavar = metavar
+        self._type = type
+
+        if default == 'NoDefault' and optional:
+            raise ValueError(f'Optional option {name} must define a default value.')
+        if default != 'NoDefault' and not optional:
+            raise ValueError(f'Required option {name} cannot take a default value.')
+
+        self.set_default(default)
+
+    @abstractmethod
+    def add_arg(self, parser: ArgumentParser) -> None: ...
+
+    @property
+    def is_optional(self) -> bool:
+        return self._optional
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def aliases(self) -> list[str]:
+        return self._aliases
+
+    @property
+    def cmd_line_name(self) -> str:
+        return self._cmd_line_name
+
+    @property
+    def toml_name(self) -> str:
+        return self._toml_name
+
+    @property
+    def help_str(self) -> str | None:
+        return self._help_str
+
+    @property
+    def metavar(self) -> str | None:
+        return self._metavar
+
+    @property
+    def default(self) -> Any:
+        return self._default
+
+    @abstractmethod
+    def set_default(self, default: Any) -> None: ...
+
+    @property
+    def type(self) -> type | UnionType:
+        return self._type
+
+
+class IntOption(Option):
+
+    def set_default(self, default: Any) -> None:
+        assert isinstance(default, int)
+        self._default = default
+
+    def add_arg(self, args: ArgumentParser) -> None:
+        args.add_argument(
+            self.cmd_line_name,
+            *(self.aliases),
+            default='NoDefault',
+            required=self.is_optional,
+            type=int,
+            help=self.help_str,
+            metavar=self.metavar,
+            dest=self.name,
+        )
+
+
+class BoolOption(Option):
```

Comment:
It'd be best to avoid the need for this class hierarchy. It is possible I'm missing something, but I think you can just add to `Option` all the attributes necessary to call `add_argument`.

Reply:
I was looking to hide the argparse implementation details, like `FileType`, but I guess this is not really necessary.